coderd/workspaceapps/db.go

```diff
@@ -341,3 +363,192 @@ func (p *DBTokenProvider) authorizeRequest(ctx context.Context, roles *rbac.Subj
 	return false, warnings, nil
 }
+
+type auditRequest struct {
+	time   time.Time
+	ip     pqtype.Inet
+	apiKey *database.APIKey
+	dbReq  *databaseRequest
+}
+
+// auditInitAutocommitRequest creates a new audit session and audit log for the
+// given request, if one does not already exist. If an audit session already
+// exists, it will be updated with the current timestamp. A session is used to
+// reduce the number of audit logs created.
+//
+// A session is unique to the agent, app, user and users IP. If any of these
+// values change, a new session and audit log is created.
+func (p *DBTokenProvider) auditInitAutocommitRequest(ctx context.Context, w http.ResponseWriter, r *http.Request) (aReq *auditRequest) {
+	// Get the status writer from the request context so we can figure
+	// out the HTTP status and autocommit the audit log.
+	sw, ok := w.(*tracing.StatusWriter)
+	if !ok {
+		panic("dev error: http.ResponseWriter is not *tracing.StatusWriter")
+	}
+
+	aReq = &auditRequest{
+		time: dbtime.Now(),
+		ip:   audit.ParseIP(r.RemoteAddr),
+	}
+
+	// Set the commit function on the status writer to create an audit
+	// log, this ensures that the status and response body are available.
+	sw.AddDoneFunc(func() {
+		if sw.Status == http.StatusSeeOther {
+			// Redirects aren't interesting as we will capture the audit
+			// log after the redirect.
+			//
+			// There's a case where we call httpmw.RedirectToLogin for
+			// path-based apps the user doesn't have access to, in which
+			// case the dashboard login redirect is used and we end up
+			// not hitting the workspaceapps API again due to dashboard
+			// showing 404. (Bug?)
+			return
+		}
+
+		if aReq.dbReq == nil {
+			// App doesn't exist, there's information in the Request
+			// struct but we need UUIDs for audit logging.
+			return
+		}
+
+		type additionalFields struct {
+			audit.AdditionalFields
+			SlugOrPort string `json:"slug_or_port,omitempty"`
+		}
+		appInfo := additionalFields{
+			AdditionalFields: audit.AdditionalFields{
+				WorkspaceOwner: aReq.dbReq.Workspace.OwnerUsername,
+				WorkspaceName:  aReq.dbReq.Workspace.Name,
+				WorkspaceID:    aReq.dbReq.Workspace.ID,
+			},
+		}
+		switch {
+		case aReq.dbReq.AccessMethod == AccessMethodTerminal:
+			appInfo.SlugOrPort = "terminal"
+		case aReq.dbReq.App.ID == uuid.Nil:
+			// If this isn't an app or a terminal, it's a port.
+			appInfo.SlugOrPort = aReq.dbReq.AppSlugOrPort
+		}
+
+		appInfoBytes, err := json.Marshal(appInfo)
+		if err != nil {
+			p.Logger.Error(ctx, "marshal additional fields failed", slog.Error(err))
+		}
+
+		userID := uuid.NullUUID{}
+		if aReq.apiKey != nil {
+			userID = uuid.NullUUID{Valid: true, UUID: aReq.apiKey.UserID}
+		}
+
+		var (
+			updatedIDs []uuid.UUID
+			sessionID  = uuid.Nil
+		)
+		err = p.Database.InTx(func(tx database.Store) error {
+			// nolint:gocritic // System context is needed to write audit sessions.
+			dangerousSystemCtx := dbauthz.AsSystemRestricted(ctx)
```

Comment:
I'd prefer for us to avoid using this system context if possible. Can we add a separate role for this?

Reply:
Understood. I don't want to make this a blocker, but at some point we need to rip off that band-aid.